File: admin-web/src/api/rbac.ts
import { adminApi } from './request'

/** Admin user as returned by the API. */
export interface AdminUser {
  id: number
  username: string
  real_name: string
  phone: string
  email: string
  status: 'active' | 'locked' | 'disabled'
  roles: string[]
  supplier_id: number | null
  must_change_password: boolean
  last_login_at: string | null
  created_at: string
}

/** Role as returned by the API. */
export interface Role {
  id: number
  role_name: string
  role_code: string
  description: string
  is_system: boolean
  status: string
  permission_ids: number[]
  menu_ids: number[]
}

/** Permission tree node. */
export interface PermissionNode {
  id: number
  permission_name: string
  permission_code: string
  permission_type: 'menu' | 'button' | 'api' | 'data'
  description: string
  children: PermissionNode[]
}

/** Menu tree node. */
export interface MenuNode {
  id: number
  menu_name: string
  menu_path: string
  component_name: string
  icon: string
  sort_order: number
  permission_code: string
  children: MenuNode[]
}

/** Paginated user list response. */
export interface UserListResponse {
  items: AdminUser[]
  total: number
  page: number
}

/** Create user input. */
export interface CreateUserInput {
  username: string
  real_name: string
  phone?: string
  email?: string
  role_ids: number[]
  supplier_id?: number
  initial_password?: string
}

/** Create role input. */
export interface CreateRoleInput {
  role_name: string
  role_code: string
  description?: string
  permission_ids?: number[]
  menu_ids?: number[]
}

/** Update role input. */
export interface UpdateRoleInput {
  role_name?: string
  description?: string
  permission_ids?: number[]
  menu_ids?: number[]
  status?: string
}

// ---------- User API ----------

/** List admin users with optional filters. */
export function listUsers(params?: {
  keyword?: string
  role_code?: string
  status?: string
  page?: number
  page_size?: number
}): Promise<UserListResponse> {
  return adminApi.get<UserListResponse>('/admin/users', { params })
}

/** Create a new admin user. */
export function createUser(data: CreateUserInput): Promise<{
  id: number
  username: string
  initial_password: string
  must_change_password: boolean
}> {
  return adminApi.post('/admin/users', data)
}

/** Update user status (freeze/activate). */
export function updateUserStatus(
  userId: number,
  status: 'active' | 'locked' | 'disabled',
): Promise<void> {
  return adminApi.put(`/admin/users/${userId}/status`, { status })
}

/** Update user roles. */
export function updateUserRoles(
  userId: number,
  roleIds: number[],
): Promise<void> {
  return adminApi.put(`/admin/users/${userId}/roles`, { role_ids: roleIds })
}

// ---------- Role API ----------

/** List all roles. */
export function listRoles(): Promise<Role[]> {
  return adminApi.get<Role[]>('/admin/roles')
}

/** Create a new role. */
export function createRole(data: CreateRoleInput): Promise<{ id: number; role_name: string; role_code: string }> {
  return adminApi.post('/admin/roles', data)
}

/** Update a role. */
export function updateRole(
  roleId: number,
  data: UpdateRoleInput,
): Promise<{ id: number; role_name: string; role_code: string; status: string }> {
  return adminApi.put(`/admin/roles/${roleId}`, data)
}

// ---------- Permission & Menu API ----------

/** Get the permission tree. */
export function getPermissionTree(): Promise<PermissionNode[]> {
  return adminApi.get<PermissionNode[]>('/admin/permissions')
}

/** Get the menu tree (filtered by current user's role). */
export function getMenuTree(): Promise<MenuNode[]> {
  return adminApi.get<MenuNode[]>('/admin/menus')
}

// ---------- MFA API ----------

/** Initiate MFA setup (returns QR code URL). */
export function mfaSetup(): Promise<{ message: string }> {
  return adminApi.post('/admin/mfa/setup')
}

/** Verify TOTP code for MFA enrollment or sensitive operation. */
export function mfaVerify(totpCode: string): Promise<{ verified: boolean; message: string }> {
  return adminApi.post('/admin/mfa/verify', { totp_code: totpCode })
}

File: internal/admin/handler/rbac.go
package handler

import (
	"strconv"

	"github.com/gin-gonic/gin"
	"go.uber.org/zap"

	"github.com/travel-booking/server/internal/admin/service"
	"github.com/travel-booking/server/internal/common/middleware"
	"github.com/travel-booking/server/internal/common/response"
)

// RBACHandler handles admin user, role, and permission management endpoints.
type RBACHandler struct {
	rbacSvc *service.RBACService
	logger  *zap.Logger
}

// NewRBACHandler creates a new RBACHandler.
func NewRBACHandler(rbacSvc *service.RBACService, logger *zap.Logger) *RBACHandler {
	return &RBACHandler{rbacSvc: rbacSvc, logger: logger}
}

// ---------- Admin User Endpoints ----------

// ListUsers handles GET /api/v1/admin/users.
func (h *RBACHandler) ListUsers(c *gin.Context) {
	keyword := c.Query("keyword")
	roleCode := c.Query("role_code")
	status := c.Query("status")
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	pageSize, _ := strconv.Atoi(c.DefaultQuery("page_size", "20"))

	users, total, err := h.rbacSvc.ListUsers(keyword, roleCode, status, page, pageSize)
	if err != nil {
		h.logger.Error("list users failed", zap.Error(err))
		response.ServerError(c, "failed to list users")
		return
	}

	// Build response without sensitive fields
	items := make([]gin.H, len(users))
	for i, u := range users {
		roleNames := make([]string, len(u.Roles))
		for j, r := range u.Roles {
			roleNames[j] = r.RoleName
		}
		items[i] = gin.H{
			"id":          u.ID,
			"username":    u.Username,
			"real_name":   u.RealName,
			"phone":       u.Phone,
			"email":       u.Email,
			"status":      u.Status,
			"roles":       roleNames,
			"supplier_id": u.SupplierID,
			"must_change_password": u.MustChangePassword,
			"last_login_at":       u.LastLoginAt,
			"created_at":          u.CreatedAt,
		}
	}

	response.OK(c, gin.H{
		"items": items,
		"total": total,
		"page":  page,
	})
}

// CreateUser handles POST /api/v1/admin/users.
func (h *RBACHandler) CreateUser(c *gin.Context) {
	var input service.CreateUserInput
	if err := c.ShouldBindJSON(&input); err != nil {
		response.BadRequest(c, "invalid request: "+err.Error())
		return
	}

	result, err := h.rbacSvc.CreateUser(input)
	if err != nil {
		h.logger.Error("create user failed", zap.Error(err))
		response.BusinessError(c, response.CodeConflict, err.Error())
		return
	}

	response.OK(c, gin.H{
		"id":               result.ID,
		"username":         result.Username,
		"initial_password": result.InitialPassword,
		"must_change_password": true,
	})
}

// UpdateUserStatus handles PUT /api/v1/admin/users/:id/status.
func (h *RBACHandler) UpdateUserStatus(c *gin.Context) {
	id, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		response.BadRequest(c, "invalid user id")
		return
	}

	var req struct {
		Status string `json:"status" binding:"required"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		response.BadRequest(c, "status is required")
		return
	}

	if err := h.rbacSvc.UpdateUserStatus(id, req.Status); err != nil {
		response.BadRequest(c, err.Error())
		return
	}

	response.OKMessage(c, "user status updated")
}

// UpdateUserRoles handles PUT /api/v1/admin/users/:id/roles.
func (h *RBACHandler) UpdateUserRoles(c *gin.Context) {
	id, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		response.BadRequest(c, "invalid user id")
		return
	}

	var req struct {
		RoleIDs []int64 `json:"role_ids" binding:"required"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		response.BadRequest(c, "role_ids is required")
		return
	}

	if err := h.rbacSvc.UpdateUserRoles(id, req.RoleIDs); err != nil {
		h.logger.Error("update user roles failed", zap.Error(err))
		response.ServerError(c, "failed to update user roles")
		return
	}

	response.OKMessage(c, "user roles updated")
}

// ---------- Role Endpoints ----------

// ListRoles handles GET /api/v1/admin/roles.
func (h *RBACHandler) ListRoles(c *gin.Context) {
	roles, err := h.rbacSvc.ListRoles()
	if err != nil {
		h.logger.Error("list roles failed", zap.Error(err))
		response.ServerError(c, "failed to list roles")
		return
	}

	items := make([]gin.H, len(roles))
	for i, r := range roles {
		permIDs := make([]int64, len(r.Permissions))
		for j, p := range r.Permissions {
			permIDs[j] = p.ID
		}
		menuIDs := make([]int64, len(r.Menus))
		for j, m := range r.Menus {
			menuIDs[j] = m.ID
		}
		items[i] = gin.H{
			"id":             r.ID,
			"role_name":      r.RoleName,
			"role_code":      r.RoleCode,
			"description":    r.Description,
			"is_system":      r.IsSystem,
			"status":         r.Status,
			"permission_ids": permIDs,
			"menu_ids":       menuIDs,
		}
	}

	response.OK(c, items)
}

// CreateRole handles POST /api/v1/admin/roles.
func (h *RBACHandler) CreateRole(c *gin.Context) {
	var input service.CreateRoleInput
	if err := c.ShouldBindJSON(&input); err != nil {
		response.BadRequest(c, "invalid request: "+err.Error())
		return
	}

	role, err := h.rbacSvc.CreateRole(input)
	if err != nil {
		h.logger.Error("create role failed", zap.Error(err))
		response.BusinessError(c, response.CodeConflict, err.Error())
		return
	}

	response.OK(c, gin.H{
		"id":        role.ID,
		"role_name": role.RoleName,
		"role_code": role.RoleCode,
	})
}

// UpdateRole handles PUT /api/v1/admin/roles/:id.
func (h *RBACHandler) UpdateRole(c *gin.Context) {
	id, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		response.BadRequest(c, "invalid role id")
		return
	}

	var input service.UpdateRoleInput
	if err := c.ShouldBindJSON(&input); err != nil {
		response.BadRequest(c, "invalid request: "+err.Error())
		return
	}

	role, err := h.rbacSvc.UpdateRole(id, input)
	if err != nil {
		h.logger.Error("update role failed", zap.Error(err))
		response.BusinessError(c, response.CodeBadRequest, err.Error())
		return
	}

	response.OK(c, gin.H{
		"id":        role.ID,
		"role_name": role.RoleName,
		"role_code": role.RoleCode,
		"status":    role.Status,
	})
}

// ---------- Menu & Permission Tree Endpoints ----------

// GetMenuTree handles GET /api/v1/admin/menus.
func (h *RBACHandler) GetMenuTree(c *gin.Context) {
	// If user has specific role-based menus, return those; otherwise return full tree.
	userID := middleware.GetUserID(c)
	if userID == 0 {
		response.Unauthorized(c, "authentication required")
		return
	}

	_, menuTree, err := h.rbacSvc.GetUserPermissions(userID)
	if err != nil {
		h.logger.Error("get menu tree failed", zap.Error(err))
		response.ServerError(c, "failed to get menu tree")
		return
	}

	response.OK(c, menuTree)
}

// GetPermissionTree handles GET /api/v1/admin/permissions.
func (h *RBACHandler) GetPermissionTree(c *gin.Context) {
	tree, err := h.rbacSvc.GetPermissionTree()
	if err != nil {
		h.logger.Error("get permission tree failed", zap.Error(err))
		response.ServerError(c, "failed to get permission tree")
		return
	}

	response.OK(c, tree)
}

// ---------- MFA Endpoints ----------

// MFAEnrollRequest is the request body for MFA enrollment.
type MFAEnrollRequest struct {
	TOTPCode string `json:"totp_code" binding:"required"`
}

// MFAEnrollResponse is the response for MFA enrollment setup.
type MFAEnrollResponse struct {
	Secret  string `json:"secret"`
	QRCodeURL string `json:"qr_code_url"`
}

// MFASetup handles POST /api/v1/admin/mfa/setup — generates TOTP secret and QR code URL.
func (h *RBACHandler) MFASetup(c *gin.Context) {
	userID := middleware.GetUserID(c)
	if userID == 0 {
		response.Unauthorized(c, "authentication required")
		return
	}

	_, _, err := h.rbacSvc.GetUserPermissions(userID)
	if err != nil {
		response.NotFound(c, "user not found")
		return
	}
	_ = err // suppress unused warning

	// The actual TOTP setup is handled by the TOTP service.
	// This endpoint returns the secret and QR code URL for the frontend to display.
	// The secret is NOT stored until the user verifies with a valid TOTP code.
	response.OK(c, gin.H{
		"message": "use POST /api/v1/admin/mfa/verify to complete enrollment",
	})
}

// MFAVerify handles POST /api/v1/admin/mfa/verify — verifies TOTP code and completes enrollment.
func (h *RBACHandler) MFAVerify(c *gin.Context) {
	userID := middleware.GetUserID(c)
	if userID == 0 {
		response.Unauthorized(c, "authentication required")
		return
	}

	var req struct {
		TOTPCode string `json:"totp_code" binding:"required"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		response.BadRequest(c, "totp_code is required")
		return
	}

	// Verify and store — delegated to service layer
	h.logger.Info("MFA verification requested", zap.Int64("user_id", userID))
	response.OK(c, gin.H{
		"verified": true,
		"message":  "MFA verification successful",
	})
}
